File: android/src/main/java/com/mealplanplus/ui/tour/Tour.kt
package com.mealplanplus.ui.tour

import androidx.compose.runtime.Composable
import androidx.compose.runtime.Stable
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.layout.boundsInRoot
import androidx.compose.ui.layout.onGloballyPositioned

/**
 * One coach-mark: highlights the on-screen element registered under [key] with a tooltip.
 * [route] (when set) is the nav destination the tour switches to before showing this step, so
 * targets that live on another screen (e.g. the More list) are composed and can report bounds.
 */
data class TourStep(
    val key: String,
    val title: String,
    val body: String,
    val route: String? = null,
)

/**
 * The guided first-run tour: walk the bottom-nav map, then drill into the More screen where
 * Foods/Meals/Diets/Groceries live. Keys are matched by [Modifier.tourTarget] call sites.
 */
val TOUR_STEPS: List<TourStep> = listOf(
    TourStep("nav_today", "Today", "Your daily hub — log meals by slot and watch your calorie ring.", route = "today"),
    TourStep("nav_plan", "Plan", "Plan diets across your week and auto-build a grocery list."),
    TourStep("nav_exercises", "Exercises", "Workouts, sessions and your exercise library."),
    TourStep("nav_health", "Health", "Track glucose, weight & blood pressure with trends."),
    TourStep("nav_misc", "More", "Foods, Meals, Diets & Groceries all live under here."),
    TourStep("misc_foods", "Foods", "Your ingredient library — add foods here.", route = "misc"),
    TourStep("misc_meals", "Meals", "Combine foods into reusable meals."),
    TourStep("misc_diets", "Diets", "Build day-plan diets from your meals, then schedule them in Plan."),
    TourStep("misc_groceries", "Groceries", "Auto-generate a shopping list from your planned diets."),
    // Closing step has no target → a centered card over the dimmed screen.
    TourStep("done", "You're all set!", "Replay this tour anytime from Settings.", route = "today"),
)

/**
 * Drives the spotlight tour: which step is active, and the live bounds of every registered target.
 * Bounds are in root coordinates so the full-screen overlay (drawn above the Scaffold, covering the
 * bottom nav) shares the same space.
 */
@Stable
class TourController {
    var running by mutableStateOf(false)
        private set
    var index by mutableStateOf(0)
        private set

    /** key → element bounds in root px, kept fresh by [Modifier.tourTarget]. */
    val targets = mutableStateMapOf<String, Rect>()

    val step: TourStep? get() = if (running) TOUR_STEPS.getOrNull(index) else null
    val isLast: Boolean get() = index >= TOUR_STEPS.lastIndex

    fun start() {
        index = 0
        running = true
    }

    fun next(onFinish: () -> Unit) {
        if (isLast) finish(onFinish) else index++
    }

    fun finish(onFinish: () -> Unit) {
        running = false
        onFinish()
    }

    fun report(key: String, bounds: Rect) {
        targets[key] = bounds
    }
}

@Composable
fun rememberTourController(): TourController = remember { TourController() }

/** Available to any screen under the nav host so it can register targets without prop-drilling. */
val LocalTourController = compositionLocalOf<TourController?> { null }

/** Registers this element as the tour target [key], reporting its position to [controller]. */
fun Modifier.tourTarget(key: String, controller: TourController?): Modifier =
    if (controller == null) this
    else this.onGloballyPositioned { controller.report(key, it.boundsInRoot()) }

/** Convenience overload that reads the controller from [LocalTourController]. */
@Composable
fun Modifier.tourTarget(key: String): Modifier = tourTarget(key, LocalTourController.current)

File: webapp/src/lib/tour.ts
import { driver, type Driver } from "driver.js";
import "driver.js/dist/driver.css";

/** Minimal shape we need from Next's router (so this stays easy to call/test). */
type RouterLike = { push: (href: string) => void };

const KEY = "mp_tour_seen";

/** Local, per-device flag — cosmetic, mirrors Android's TourStore (a fresh browser re-shows it). */
export function hasSeenTour(): boolean {
  try { return localStorage.getItem(KEY) === "1"; } catch { return true; }
}
export function markTourSeen(): void {
  try { localStorage.setItem(KEY, "1"); } catch { /* ignore */ }
}
export function resetTour(): void {
  try { localStorage.removeItem(KEY); } catch { /* ignore */ }
}

/** Resolve once the selector exists in the DOM (or after [timeout] ms), so we can wait out route
 *  changes before the spotlight anchors to a freshly-rendered element. */
function waitForEl(selector: string, timeout = 3000): Promise<Element | null> {
  return new Promise((resolve) => {
    const start = Date.now();
    const tick = () => {
      const el = document.querySelector(selector);
      if (el) return resolve(el);
      if (Date.now() - start > timeout) return resolve(null);
      requestAnimationFrame(tick);
    };
    tick();
  });
}

// Mirrors the Android tour: the 5 bottom-nav tabs, then drill into the More library, then a
// closing card. Nav tabs are on every page; the misc-* targets only exist on /misc.
function buildSteps() {
  const top = { side: "top" as const, align: "center" as const };
  const drill = { side: "bottom" as const, align: "start" as const };
  return [
    { element: '[data-tour="nav-today"]', popover: { title: "Today", description: "Your daily hub — log meals by slot and watch your calorie ring.", ...top } },
    { element: '[data-tour="nav-plan"]', popover: { title: "Plan", description: "Plan diets across your week and auto-build a grocery list.", ...top } },
    { element: '[data-tour="nav-exercises"]', popover: { title: "Exercises", description: "Workouts, sessions and your exercise library.", ...top } },
    { element: '[data-tour="nav-health"]', popover: { title: "Health", description: "Track glucose, weight & blood pressure with trends.", ...top } },
    { element: '[data-tour="nav-misc"]', popover: { title: "More", description: "Foods, Meals, Diets & Groceries all live under here.", ...top } },
    { element: '[data-tour="misc-foods"]', popover: { title: "Foods", description: "Your ingredient library — add foods here.", ...drill } },
    { element: '[data-tour="misc-meals"]', popover: { title: "Meals", description: "Combine foods into reusable meals.", ...drill } },
    { element: '[data-tour="misc-diets"]', popover: { title: "Diets", description: "Build day-plan diets from your meals, then schedule them in Plan.", ...drill } },
    { element: '[data-tour="misc-groceries"]', popover: { title: "Groceries", description: "Auto-generate a shopping list from your planned diets.", side: "top" as const, align: "start" as const } },
    { popover: { title: "You're all set!", description: "Replay this tour anytime from Settings." } },
  ];
}

// After the "More" nav step (index 4) the tour jumps to the /misc page for the drill-in steps.
const MISC_ENTRY_INDEX = 4;

/** Start the guided spotlight tour. Navigates to Today first (where the nav is), then drives
 *  Driver.js; between the nav map and the drill-in it routes to /misc and waits for it to render. */
export async function startTour(router: RouterLike): Promise<void> {
  markTourSeen(); // one-shot: this run counts as "seen" so it won't auto-open again
  if (typeof window !== "undefined" && window.location.pathname !== "/today") {
    router.push("/today");
  }
  await waitForEl('[data-tour="nav-today"]');

  const d: Driver = driver({
    showProgress: true,
    showButtons: ["next", "close"],
    popoverClass: "mp-tour",
    steps: buildSteps(),
    onNextClick: async () => {
      if (d.getActiveIndex() === MISC_ENTRY_INDEX) {
        router.push("/misc");
        await waitForEl('[data-tour="misc-foods"]');
      }
      d.moveNext();
    },
    onCloseClick: () => { d.destroy(); },
  });
  d.drive();
}
